File: apps/desktop/src/services/cache.ts
/**
 * What this device already knows, so opening the app is not a spinner.
 *
 * The port of `apps/android/core/src/main/java/com/aktech/nexora/core/store/Cache.kt`,
 * and it keeps that file's one rule: **nothing in here is plaintext**. Messages
 * are stored exactly as the server sent them - a sealed envelope this client
 * happens to hold the key for - and the lists are things the server would hand
 * to anyone who asked with this account's token. Writing decrypted bodies to
 * disk would undo what the encryption is for, which is why the decrypted
 * history in `stores/chat.ts` stays in memory and dies with the window.
 *
 * A cache has nothing in it that cannot be fetched again, so every failure path
 * here is "carry on without it". Nothing throws into a caller.
 */
import type { Channel, DirectChannel, Message, ServerWithRole } from '@nexora/shared-types';

const DB_NAME = 'nexora-cache';
const DB_VERSION = 1;
const LISTS = 'lists';
const MESSAGES = 'messages';
const OWNER_KEY = 'owner';

/** Messages kept per channel. Fifty is the server's page; this holds ten of them. */
const MESSAGES_PER_CHANNEL = 500;

let open: Promise<IDBDatabase | null> | null = null;

function database(): Promise<IDBDatabase | null> {
  if (open) return open;

  open = new Promise((resolve) => {
    if (typeof indexedDB === 'undefined') {
      resolve(null);
      return;
    }
    const request = indexedDB.open(DB_NAME, DB_VERSION);

    request.onupgradeneeded = () => {
      const db = request.result;
      if (!db.objectStoreNames.contains(LISTS)) db.createObjectStore(LISTS);
      if (!db.objectStoreNames.contains(MESSAGES)) {
        const store = db.createObjectStore(MESSAGES, { keyPath: 'id' });
        // Read by channel, ordered by time - which is the only read there is.
        store.createIndex('channel', ['channelId', 'createdAt']);
      }
    };
    request.onsuccess = () => resolve(request.result);
    // A browser with storage denied, a private window, a corrupt database: all
    // of them are an app that works and does not open instantly.
    request.onerror = () => resolve(null);
  });

  return open;
}

function done(transaction: IDBTransaction): Promise<void> {
  return new Promise((resolve) => {
    transaction.oncomplete = () => resolve();
    transaction.onerror = () => resolve();
    transaction.onabort = () => resolve();
  });
}

function ask<T>(request: IDBRequest<T>): Promise<T | null> {
  return new Promise((resolve) => {
    request.onsuccess = () => resolve(request.result);
    request.onerror = () => resolve(null);
  });
}

// --- Lists ------------------------------------------------------------------

async function readList<T>(key: string): Promise<T | null> {
  const db = await database();
  if (!db) return null;
  const store = db.transaction(LISTS, 'readonly').objectStore(LISTS);
  return (await ask<T>(store.get(key) as IDBRequest<T>)) ?? null;
}

async function writeList(key: string, value: unknown): Promise<void> {
  const db = await database();
  if (!db) return;
  const transaction = db.transaction(LISTS, 'readwrite');
  transaction.objectStore(LISTS).put(value, key);
  await done(transaction);
}

export const cache = {
  /**
   * Binds the cache to an account. A different one than last time means
   * everything here belongs to somebody else and goes.
   *
   * Called before either store reads from it, so a cache belonging to another
   * account is gone rather than merely about to be.
   */
  async claim(userId: string): Promise<void> {
    const owner = await readList<string>(OWNER_KEY);
    if (owner === userId) return;
    await cache.clear();
    await writeList(OWNER_KEY, userId);
  },

  /**
   * Forgets everything. Signing out is the one moment somebody has said they
   * are done with this device - a session that merely expired keeps its cache,
   * so signing back in is still instant.
   */
  async clear(): Promise<void> {
    const db = await database();
    if (!db) return;
    const transaction = db.transaction([LISTS, MESSAGES], 'readwrite');
    transaction.objectStore(LISTS).clear();
    transaction.objectStore(MESSAGES).clear();
    await done(transaction);
  },

  servers: (): Promise<ServerWithRole[] | null> => readList('servers'),
  putServers: (servers: ServerWithRole[]): Promise<void> => writeList('servers', servers),

  channels: (serverId: string): Promise<Channel[] | null> => readList(`channels:${serverId}`),
  putChannels: (serverId: string, channels: Channel[]): Promise<void> =>
    writeList(`channels:${serverId}`, channels),

  directs: (): Promise<DirectChannel[] | null> => readList('directs'),
  putDirects: (directs: DirectChannel[]): Promise<void> => writeList('directs', directs),

  /**
   * Where each channel had been read up to. Cached because the unread line is
   * drawn from it the moment a channel opens, and on a cold start that is well
   * before the network has answered - which is how a restart used to lose the
   * line entirely.
   */
  readMarkers: (): Promise<Record<string, string | null> | null> => readList('readMarkers'),
  putReadMarkers: (markers: Record<string, string | null>): Promise<void> =>
    writeList('readMarkers', markers),

  // --- Messages -------------------------------------------------------------

  /** The newest page of a channel, oldest first, as the screen wants it. */
  async messages(channelId: string, limit = 50): Promise<Message[]> {
    const db = await database();
    if (!db) return [];

    const index = db.transaction(MESSAGES, 'readonly').objectStore(MESSAGES).index('channel');
    const range = IDBKeyRange.bound([channelId, ''], [channelId, '￿']);
    const newest: Message[] = [];

    await new Promise<void>((resolve) => {
      // Backwards from the newest, because "the last fifty" is the question -
      // walking forwards would read the whole channel to answer it.
      const cursor = index.openCursor(range, 'prev');
      cursor.onsuccess = () => {
        const at = cursor.result;
        if (!at || newest.length >= limit) {
          resolve();
          return;
        }
        newest.push(at.value as Message);
        at.continue();
      };
      cursor.onerror = () => resolve();
    });

    return newest.reverse();
  },

  /**
   * Keeps the last few hundred per channel. History fetched once should not
   * need fetching again, and a few thousand rows of ciphertext is not a size
   * worth worrying about - but a channel read back to its first message is,
   * so the oldest are dropped rather than kept forever.
   */
  async putMessages(messages: Message[]): Promise<void> {
    if (messages.length === 0) return;
    const db = await database();
    if (!db) return;

    const transaction = db.transaction(MESSAGES, 'readwrite');
    const store = transaction.objectStore(MESSAGES);
    for (const message of messages) store.put(message);
    await done(transaction);

    for (const channelId of new Set(messages.map((message) => message.channelId))) {
      await prune(channelId);
    }
  },
};

/** Drops everything past the ceiling in one channel, oldest first. */
async function prune(channelId: string): Promise<void> {
  const db = await database();
  if (!db) return;

  const transaction = db.transaction(MESSAGES, 'readwrite');
  const index = transaction.objectStore(MESSAGES).index('channel');
  const range = IDBKeyRange.bound([channelId, ''], [channelId, '￿']);
  let seen = 0;

  await new Promise<void>((resolve) => {
    const cursor = index.openCursor(range, 'prev');
    cursor.onsuccess = () => {
      const at = cursor.result;
      if (!at) {
        resolve();
        return;
      }
      seen += 1;
      if (seen > MESSAGES_PER_CHANNEL) at.delete();
      at.continue();
    };
    cursor.onerror = () => resolve();
  });

  await done(transaction);
}

File: apps/android/core/src/main/java/com/aatech/betweenus/core/store/Cache.kt
package com.aatech.betweenus.core.store

import android.content.Context
import androidx.room.Dao
import androidx.room.Database
import androidx.room.Entity
import androidx.room.Index
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.PrimaryKey
import androidx.room.Query
import androidx.room.Room
import androidx.room.RoomDatabase
import com.aatech.betweenus.core.data.Channel
import com.aatech.betweenus.core.data.DirectChannel
import com.aatech.betweenus.core.data.Friend
import com.aatech.betweenus.core.data.Message
import com.aatech.betweenus.core.data.ServerMember
import com.aatech.betweenus.core.data.ServerWithRole
import com.aatech.betweenus.core.data.jsonArrayOfObjects
import com.aatech.betweenus.core.data.map
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import org.json.JSONArray
import org.json.JSONObject

/**
 * A row of whatever list it is named after, as the JSON the server sent.
 *
 * Storing the wire shape rather than a column per field is deliberate. These
 * lists are read and written whole, never queried into, and `Models.kt` already
 * holds a parser for every one of them - a column layout would buy nothing and
 * would have to be migrated every time a field is added to the contract.
 */
@Entity(tableName = "cache")
internal data class CacheRow(@PrimaryKey val id: String, val json: String)

/**
 * One message, still sealed. `channelId` and `createdAt` are columns because
 * they are what the reads sort and filter on; the rest stays in the blob.
 */
@Entity(tableName = "message", indices = [Index("channelId")])
internal data class MessageRow(
    @PrimaryKey val id: String,
    val channelId: String,
    val createdAt: String,
    val json: String,
)

@Dao
internal interface CacheDao {
    @Query("SELECT json FROM cache WHERE id = :id")
    suspend fun read(id: String): String?

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun write(row: CacheRow)

    @Query("DELETE FROM cache")
    suspend fun clearLists()

    /** Newest first, which is the cheap end of the index to read a page from. */
    @Query("SELECT * FROM message WHERE channelId = :channelId ORDER BY createdAt DESC LIMIT :limit")
    suspend fun latest(channelId: String, limit: Int): List<MessageRow>

    @Query(
        "SELECT * FROM message WHERE channelId = :channelId AND createdAt < :before " +
            "ORDER BY createdAt DESC LIMIT :limit",
    )
    suspend fun before(channelId: String, before: String, limit: Int): List<MessageRow>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun writeMessages(rows: List<MessageRow>)

    @Query("DELETE FROM message")
    suspend fun clearMessages()

    @Query("DELETE FROM message WHERE channelId = :channelId")
    suspend fun clearChannel(channelId: String)

    @Query("DELETE FROM message WHERE id IN (:ids)")
    suspend fun forgetMessages(ids: List<String>)
}

@Database(entities = [CacheRow::class, MessageRow::class], version = 1, exportSchema = false)
internal abstract class BetweenUsDatabase : RoomDatabase() {
    abstract fun cache(): CacheDao
}

/**
 * What the app knew last time it ran.
 *
 * The point of it is that opening BetweenUs shows a conversation rather than a
 * spinner: the stores draw from here first and let the network catch up behind
 * them, the way WhatsApp does. Nothing here is authoritative - the server is,
 * and every read is followed by a refresh - but on a cold start, on a train,
 * or against a backend that is having a bad day, this is what is on screen.
 *
 * Two things are deliberately *not* stored:
 *
 * - **Plaintext.** Messages are cached exactly as they arrived, still sealed.
 *   Decryption happens on the way to the screen. A database pulled off a rooted
 *   phone is worth what the server's own rows are worth, which is nothing
 *   without a channel key - and those live in the Keystore, not here.
 * - **Anything for another account.** [claim] wipes the lot when the user id
 *   changes, and every read waits for it, so a second account cannot flash the
 *   first one's servers on screen while the network catches up.
 *
 * Writes are fire-and-forget: a cache that cannot be written is a slower app,
 * never a broken one, so nothing here throws into a caller.
 */
object Cache {
    private const val OWNER = "owner"
    private const val PAGE = 50

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    @Volatile
    private var db: BetweenUsDatabase? = null

    /** Completes when the cache is known to belong to whoever is signed in. */
    @Volatile
    private var claimed = CompletableDeferred<Unit>()

    fun init(context: Context) {
        db = Room.databaseBuilder(
            context.applicationContext,
            BetweenUsDatabase::class.java,
            "betweenus-cache.db",
        )
            // A cache has nothing in it that cannot be fetched again, so a
            // schema change throws it away rather than carrying a migration.
            .fallbackToDestructiveMigration(dropAllTables = true)
            .build()
    }

    /**
     * Binds the cache to an account. A different one than last time means
     * everything here belongs to somebody else and goes.
     */
    fun claim(userId: String) {
        scope.launch {
            runCatching {
                val dao = db?.cache() ?: return@runCatching
                if (dao.read(OWNER) != userId) {
                    dao.clearLists()
                    dao.clearMessages()
                    dao.write(CacheRow(OWNER, userId))
                }
            }
            claimed.complete(Unit)
        }
    }

    /**
     * Forgets everything. Signing out is the one moment a person has said they
     * are done with this device - a session that merely expired keeps its cache,
     * so signing back in is still instant.
     */
    suspend fun clear() {
        claimed = CompletableDeferred()
        runCatching {
            db?.cache()?.let {
                it.clearLists()
                it.clearMessages()
            }
        }
    }

    /**
     * Forgets one conversation's messages and nothing else.
     *
     * What "clear this chat" needs. [clear] would work and is shorter, but it
     * throws away every other conversation this phone has cached - so clearing
     * one chat would turn the next four things somebody opens into spinners,
     * for a reason they could not connect to what they just did.
     *
     * The lists are left alone: the channel still exists and still belongs in
     * the rail. What went is what was said in it.
     */
    suspend fun forgetChannel(channelId: String) {
        runCatching { db?.cache()?.clearChannel(channelId) }
    }

    /**
     * Forgets specific messages, by id.
     *
     * For the two deletions that leave no tombstone - a one-time message that
     * was opened, and one whose disappearing window closed. An ordinary delete
     * does not come through here: its tombstone is written over the old row by
     * [putMessages], which is what makes a conversation still say "this was
     * here and is gone" after the app is restarted.
     */
    suspend fun forgetMessages(ids: List<String>) {
        if (ids.isEmpty()) return
        runCatching { db?.cache()?.forgetMessages(ids) }
    }

    // --- lists ---

    suspend fun servers(): List<ServerWithRole>? = readList("servers") { ServerWithRole.from(it) }

    fun putServers(servers: List<ServerWithRole>) =
        writeList("servers", jsonArrayOfObjects(servers) { it.toJson() })

    suspend fun channels(): Map<String, List<Channel>>? =
        readMap("channels") { array -> array.map { Channel.from(it) } }

    fun putChannels(channels: Map<String, List<Channel>>) =
        writeMap("channels", channels.mapValues { (_, list) -> jsonArrayOfObjects(list) { it.toJson() } })

    suspend fun directChannels(): List<DirectChannel>? =
        readList("directChannels") { DirectChannel.from(it) }

    fun putDirectChannels(channels: List<DirectChannel>) =
        writeList("directChannels", jsonArrayOfObjects(channels) { it.toJson() })

    suspend fun friends(): List<Friend>? = readList("friends") { Friend.from(it) }

    fun putFriends(friends: List<Friend>) =
        writeList("friends", jsonArrayOfObjects(friends) { it.toJson() })

    suspend fun members(): Map<String, List<ServerMember>>? =
        readMap("members") { array -> array.map { ServerMember.from(it) } }

    fun putMembers(members: Map<String, List<ServerMember>>) =
        writeMap("members", members.mapValues { (_, list) -> jsonArrayOfObjects(list) { it.toJson() } })

    suspend fun unread(): Map<String, Int>? = read("unread")?.let { stored ->
        runCatching {
            val json = JSONObject(stored)
            json.keys().asSequence().associateWith { json.getInt(it) }
        }.getOrNull()
    }

    fun putUnread(unread: Map<String, Int>) {
        val json = JSONObject()
        unread.forEach { (channelId, count) -> json.put(channelId, count) }
        write("unread", json.toString())
    }

    // --- messages ---

    /** The newest page of a channel, oldest first, as the screen wants it. */
    suspend fun messages(channelId: String): List<Message> =
        readMessages { it.latest(channelId, PAGE) }

    /** The page before [before], which is a `createdAt`, not an id. */
    suspend fun messagesBefore(channelId: String, before: String): List<Message> =
        readMessages { it.before(channelId, before, PAGE) }

    /**
     * Keeps everything, and is meant to: history that has been fetched once
     * should never need fetching again, and a few thousand rows of text is not
     * a size worth pruning. If it ever becomes one, prune by channel here.
     */
    fun putMessages(messages: List<Message>) {
        if (messages.isEmpty()) return
        scope.launch {
            runCatching {
                dao()?.writeMessages(
                    messages.map {
                        MessageRow(it.id, it.channelId, it.createdAt, it.toJson().toString())
                    },
                )
            }
        }
    }

    // --- plumbing ---

    private suspend fun dao(): CacheDao? {
        claimed.await()
        return db?.cache()
    }

    private suspend fun readMessages(query: suspend (CacheDao) -> List<MessageRow>): List<Message> =
        runCatching {
            val dao = dao() ?: return emptyList()
            query(dao).mapNotNull { row ->
                runCatching { Message.from(JSONObject(row.json)) }.getOrNull()
            }.sortedBy { it.createdAt }
        }.getOrDefault(emptyList())

    private suspend fun read(id: String): String? =
        runCatching { dao()?.read(id) }.getOrNull()

    private fun write(id: String, json: String) {
        scope.launch { runCatching { dao()?.write(CacheRow(id, json)) } }
    }

    private suspend fun <T> readList(id: String, parse: (JSONObject) -> T): List<T>? =
        read(id)?.let { stored -> runCatching { JSONArray(stored).map(parse) }.getOrNull() }

    private fun writeList(id: String, array: JSONArray) = write(id, array.toString())

    private suspend fun <T> readMap(id: String, parse: (JSONArray) -> List<T>): Map<String, List<T>>? =
        read(id)?.let { stored ->
            runCatching {
                val json = JSONObject(stored)
                json.keys().asSequence().associateWith { parse(json.getJSONArray(it)) }
            }.getOrNull()
        }

    private fun writeMap(id: String, byKey: Map<String, JSONArray>) {
        val json = JSONObject()
        byKey.forEach { (key, array) -> json.put(key, array) }
        write(id, json.toString())
    }
}
